CORE-2157: cloud_storage_clients: add support for path-style addressing

WillemKauf opened the pull request:
Fixes #2183.

This PR adds support for path-style requests to S3 cloud storage.

`cloud_storage_url_style` is a configurable option in `redpanda`. It defaults to the existing `virtual_host` style, but can be set to `path`. The `s3_client` will now have its headers for requests generated based on the configured URL style.

Support for ducktape has been included, and path-style requests have been verified as working using the `s3_test_client_main` test. Ideally, the cloud storage self-test in PR #17586 would be integrated with these changes so as to verify both virtual-hosted style requests and path style requests.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.3.x
- [ ] v23.2.x

## Release Notes

### Features

* Adds path-style addressing to S3 client requests. 
* This option is configurable through `rpk cluster config edit`:

```
# The addressing style to use for S3 requests. (one of virtual_host, path, restart required)
cloud_storage_url_style: path
```

Lazin commented on src/v/cloud_storage_clients/s3_url_style.h:
```diff
@@ -0,0 +1,28 @@
+/*
```
I think stuff like this could go into the `cloud_storage_clients/types.h`.

WillemKauf replied:
Agreed. Fixed!

Lazin commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -343,6 +346,30 @@ request_creator::make_delete_objects_request(
+        return fmt::format("{}.{}", name(), _ap());
+    case s3_url_style::path:
+        // Host: s3.region-code.amazonaws.com
+        return fmt::format("{}", _ap());
```
why the bucket name is not included?

WillemKauf replied:
Following the [S3 User Guide's examples of path-style requests](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html#VirtualHostingExamples), this seems to be the proper form for the host in the header.

The bucket name is instead used in the target of the header.

In commit 3d4ef12e2c4b7d69ddec9a838e9114bf8372be47, I added some comments in the various `request_creator::make_*()` functions to reflect this style.

dotnwat replied:
> Following the [S3 User Guide's examples of path-style requests](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html#VirtualHostingExamples), this seems to be the proper form for the host in the header.
> 
> The bucket name is instead used in the target of the header.
> 
> In commit [8edf7d5](https://github.com/redpanda-data/redpanda/commit/8edf7d571c2050cdf212135a65020f52d055655e), I added some comments in the various `request_creator::make_*()` functions to reflect this style.

I was also confused here. But the answer is that this is `make_host`, not `make_target`, right?

WillemKauf replied:
Yes, correct. This is `make_host`.

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/47875#018ee7f0-c1eb-450a-9157-d56b4c199846:
```
"rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=node_add.test_mode=TestMode.NO_TIRED_STORAGE.cleanup_policy=compact"
"rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=off.test_mode=TestMode.FAST_MOVES.cleanup_policy=compact"
"rptest.tests.e2e_shadow_indexing_test.ShadowIndexingInfiniteRetentionTest.test_segments_not_deleted.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=off.test_mode=TestMode.TIRED_STORAGE.cleanup_policy=compact"
"rptest.tests.retention_policy_test.ShadowIndexingCloudRetentionTest.test_topic_recovery_retention_settings"
"rptest.tests.retention_policy_test.ShadowIndexingLocalRetentionTest.test_local_time_based_retention_is_overridden.local_retention_ms=3600000.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.retention_policy_test.ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention.cluster_remote_write=False.topic_remote_write=false.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.retention_policy_test.ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention.cluster_remote_write=True.topic_remote_write=-1.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.retention_policy_test.ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention.cluster_remote_write=True.topic_remote_write=true.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.topic_creation_test.CreateSITopicsTest.test_shadow_indexing_mode_persistence"
"rptest.tests.archive_retention_test.CloudArchiveRetentionTest.test_delete.cloud_storage_type=CloudStorageType.S3.retention_type=retention.bytes"
"rptest.tests.e2e_iam_role_test.AWSRoleFetchTests.test_write"
"rptest.tests.cloud_storage_chunk_read_path_test.CloudStorageChunkReadTest.test_fallback_mode"
"rptest.tests.cloud_storage_chunk_read_path_test.CloudStorageChunkReadTest.test_read_chunks"
"rptest.tests.read_replica_e2e_test.TestReadReplicaService.test_identical_lwms_after_delete_records.partition_count=5.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.read_replica_e2e_test.TestReadReplicaService.test_writes_forbidden.partition_count=10.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.cluster_recovery_test.ClusterRecoveryTest.test_basic_controller_snapshot_restore"
"rptest.tests.offset_for_leader_epoch_read_replica_test.OffsetForLeaderEpochReadReplicaTest.test_offset_for_leader_epoch"
"rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTest.test_write.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.follower_fetching_test.FollowerFetchingTest.test_basic_follower_fetching.read_from_object_store=True"
"rptest.tests.recovery_mode_test.RecoveryModeTest.test_rolling_restart"
"rptest.tests.e2e_shadow_indexing_test.EndToEndHydrationTimeoutTest.test_hydration_completes_when_consumer_killed"
"rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTest.test_reset_from_cloud.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.e2e_shadow_indexing_test.EndToEndSpilloverTest.test_spillover.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.full_disk_test.LogStorageMaxSizeSI.test_stay_below_target_size.log_segment_size=1048576.cleanup_policy=compact"
"rptest.tests.idempotency_test.IdempotencySnapshotDelivery.test_recovery_after_snapshot_is_delivered"
"rptest.tests.shadow_indexing_compacted_topic_test.ShadowIndexingCompactedTopicTest.test_upload.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type=CloudStorageType.S3.test_case=.TS_Read==True.AdjacentSegmentMergerReupload==True.SpilloverManifestUploaded==True"
"rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type=CloudStorageType.S3.test_case=.TS_Read==True.TS_Timequery==True"
"rptest.tests.topic_delete_test.TopicDeleteCloudStorageTest.drop_lifecycle_marker_test.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_fast2.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_many_partitions.cloud_storage_type=CloudStorageType.S3.check_mode=no_check"
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_missing_segment.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.adjacent_segment_merging_test.AdjacentSegmentMergingTest.test_reupload_of_local_segments.acks=1.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.cluster_metadata_upload_test.ClusterMetadataUploadTest.test_uploads_after_wipe.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_segment_deletion.cloud_storage_enabled=True.truncate_point=one_below_high_watermark"
"rptest.tests.license_upgrade_test.UpgradeMigratingLicenseVersion.test_license_upgrade.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.shadow_indexing_admin_api_test.SIAdminApiTest.test_topic_recovery_redirects_to_controller_leader"
"rptest.tests.topic_creation_test.CreateTopicsTest.test_no_log_bloat_when_recreating_existing_topics"
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_empty_segments.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_no_data.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_size_based_retention.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_vcluster_id.cloud_storage_type=CloudStorageType.S3"
```
<!--1cb99842-28fe-45ac-9466-cebf4bf5ca67-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/47883#018ee878-3cb1-4e71-ba4b-b8f2ea993e47
<!--2b5b34fe-3adb-4df4-b924-36962d809150-->

Lazin commented on src/v/cloud_storage_clients/test_client/s3_test_client_main.cc:
```diff
@@ -76,6 +77,11 @@ void cli_opts(boost::program_options::options_description_easy_init opt) {
 
+    opt(
+      "url_style",
+      po::value<std::string>()->default_value("virtual_host"),
```
This is good for development but we also need a proper unit-test. In the `src/v/cloud_storage/remote.h` we have a `remote` class which serves as a proxy for the code that has to interact with the cloud storage. All code paths which are downloading or uploading or deleting are using this class instead of the client. The unit-test has to show that when the path-style is enabled no REST API request uses virtual host style URL. The `cloud_storage::remote` has a unit-test suite. It invokes every method of the `remote` and checks the results using the `s3_imposter`. The imposter allows you to examine parameters of the request and also validate the URL. So basically, we need to have all these tests to be invoked with virtual host style URLs and also with path-style URLs.

The motivation here is that some cloud storage providers may not support one style or the other. In this situation we don't want to accidentally use the wrong style.

Lazin replied:
Looks like all existing REST API's that we're using are supported. But we can add new request in the future so some check is nice to have.

Lazin replied:
Could be done in a followup.

dotnwat replied:
> Could be done in a followup.

@Lazin Yes, and we are also going to wire up ducktape wiht minio running in both modes (pretty sure it supports it). You can see all teh follow up work in CORE-725.

@WillemKauf can you capture Evgeny's suggestions about additional unit testing in a new or existing jira ticket as a subtask for the overall project?

Lazin reviewed (approved):
LGTM

WillemKauf commented:
Removed a commit that attempted to add virtual-host style configuration to the [`boto3.client`](https://botocore.amazonaws.com/v1/documentation/api/latest/reference/config.html).

Contrary to what the understanding was, Minio by default works with path style requests, [unless `MINIO_DOMAIN` is set](https://min.io/docs/minio/linux/reference/minio-server/settings/core.html#envvar.MINIO_DOMAIN). 

This environmental variable appears to be set in `docker/docker_compose.yml`, but ducktape tests utilizing the `boto3.client` are still failing to make requests with the virtual-host style to minIO. I have removed the broken code for now, and will follow up in a future PR to add virtual-host support to minIO and `boto3.client` in our ducktape tests once I figure out why it is failing.

dotnwat commented on src/v/cloud_storage_clients/types.h:
```diff
@@ -62,6 +63,17 @@ inline std::error_code make_error_code(error_outcome e) noexcept {
+enum class s3_url_style { virtual_host = 0, path };
+
+inline std::ostream& operator<<(std::ostream& os, const s3_url_style& us) {
+    switch (us) {
+    case s3_url_style::virtual_host:
+        return os << "virtual_host";
+    case s3_url_style::path:
+        return os << "path";
+    }
```
can you stick the implementation in a .cc file?

WillemKauf replied:
It's a bit tricky, because the `config` library is going to depend on this implementation. Having it `inline` and accessible from the header saves us from having to depend on the `cloud_storage_clients` library (and adding this would create a cyclic dependency).

dotnwat replied:
ahh, yeh, then don't worry about it. we'll fix those dependencies as we are switching modules over to be new-style. it's a lot of work no need to do that here.

dotnwat commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -343,6 +346,30 @@ request_creator::make_delete_objects_request(
 }
 
+std::string
+request_creator::make_host([[maybe_unused]] const bucket_name& name) const {
```
you do don't need `maybe_unused` -- it's only needed if the code generated doesn't use it (e.g. in conjunction with constexpr or templates etc...). in this case it is used in the virtual host case.

dotnwat commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -343,6 +346,30 @@ request_creator::make_delete_objects_request(
+}
+
+std::string request_creator::make_target(
+  [[maybe_unused]] const bucket_name& name, const object_key& key) const {
```
you can drop maybe_unused.

dotnwat commented on src/v/cloud_storage_clients/s3_client.h:
```diff
@@ -106,7 +107,14 @@ class request_creator {
       std::optional<char> delimiter = std::nullopt);
 
 private:
+    std::string make_host([[maybe_unused]] const bucket_name& name) const;
+
+    std::string make_target(
+      [[maybe_unused]] const bucket_name& name, const object_key& key) const;
```
you can drop maybe_unused.

dotnwat commented on src/v/cloud_storage_clients/s3_client.h:
```diff
@@ -106,7 +107,14 @@ class request_creator {
       std::optional<char> delimiter = std::nullopt);
 
 private:
+    std::string make_host([[maybe_unused]] const bucket_name& name) const;
+
+    std::string make_target(
+      [[maybe_unused]] const bucket_name& name, const object_key& key) const;
```
ss::sstring ?

WillemKauf replied:
`Boost.Beast` expects an `std::string_view` in many places, such as `header.insert()` or `header.target()`, which is why this returns an `std::string` instead of an `ss::sstring`. 

I can refactor it to return an `ss::sstring` and pass it's data and length to `Boost.Beast` functions that expect a `string_view`, or it can remain an `std::string`- what do you think?

dotnwat replied:
let's leave it as std::string. thanks for the explantion.

dotnwat commented on src/v/config/configuration.cc:
```diff
@@ -1630,6 +1630,18 @@ configuration::configuration()
       {.visibility = visibility::user},
       std::nullopt,
       &validate_non_empty_string_opt)
+  , cloud_storage_url_style(
+      *this,
+      "cloud_storage_url_style",
+      "The addressing style to use for S3 requests.",
+      {.needs_restart = needs_restart::yes,
+       .example = "virtual_host",
+       .visibility = visibility::user},
+      cloud_storage_clients::s3_url_style::virtual_host,
+      {
+        cloud_storage_clients::s3_url_style::virtual_host,
+        cloud_storage_clients::s3_url_style::path,
+      })
```
Should we add an `auto` option? From what I can tell the strategy is basically to try virtual-host and then fallback to path-style, and cache the decision.

See https://github.com/boto/botocore/blob/develop/botocore/utils.py#L1375-L1381

@Lazin what do you think?

@WillemKauf if we all think this is a good idea, let's create a new jira ticket to capture it--no need to expand the scope of this PR.

WillemKauf replied:
I think this is a great idea for future work! +1

dotnwat commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -264,8 +289,13 @@ request_creator::make_delete_objects_request(
     // will generate this request:
     //
+    // Virtual Style:
     // POST /?delete HTTP/1.1
-    // Host: <Bucket>.s3.amazonaws.com
+    // Host: {bucket-name}.s3.{region}.amazonaws.com
+    // Path Style:
+    // POST /{bucket-name}/?delete HTTP/1.1
+    // Host: s3.{region}.amazonaws.com
+    //
```
Thanks for the comments!

dotnwat reviewed (commented):
This looks great. There are a few small nits that looks like they could be a commit tacked onto the end. Then I think we should approve this, merge it right after we branch 24.1, and move on to the other tasks in CORE-725 like expanded testing.

BTW, please update this PR so that it isn't setup to close CORE-725 when it merges, which is the high-level task. Instead, it looks like CORE-2157 is the right one for this PR. Wdyt?

dotnwat reviewed (dismissed):
Approved -- let's just wait a couple days for 24.1 branching before merging.

dotnwat commented:
/ci-repeat 1

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/48181#018f0c84-6c7f-4a56-ae78-5ca002edee99:
```
"rptest.tests.upgrade_test.UpgradeFromPriorFeatureVersionCloudStorageTest.test_rolling_upgrade.cloud_storage_type=CloudStorageType.S3"
```
<!--ce8fca78-9fba-4fa1-b52c-59e6f75bb5b8-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/48181#018f0c8a-f917-47dc-9d33-9c5d0e34f7b9
<!--aa505702-a795-4841-9bb5-db83fa67bc80-->

micheleRP reviewed (approved):
lgtm from docs side!

dotnwat reviewed (changes requested):
changes look great, we should merge this as soon as we take care of a what looks like some related failures.

looks like a few ducktape tests are complaining about a new configuration option.

```
INFO  2024-04-23 20:34:59,789 [shard 0] redpanda::main - application.cc:269 - Failure during startup: std::invalid_argument (Unknown property cloud_storage_url_style)
```

dotnwat reviewed (approved):
Failures are from the 24.1 branching issue in upgrade tests, and https://github.com/redpanda-data/redpanda/issues/17920.
